### PR TITLE
476 Add type hints to monai/data/

### PR DESCRIPTION
**Pull request by benjamin-gorman**
### Description

Iteration of #476

- remove unused imports
- reorder imports
- add type hints
- fix type hints
- remove docstring type hints

### Status

**Ready**

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Docstrings/Documentation updated

### CODE REVIEW

**Comment by wyli** [comment]
could you help check this PR @Nic-Ma, for folder path string types I'm not 100% sure

File: monai/data/dataset.py [hunk 1]
```diff
@@ -38,7 +39,7 @@ class Dataset(_TorchDataset):
          },                           },                           }]
     """
 
-    def __init__(self, data, transform: Optional[Callable] = None):
+    def __init__(self, data, transform: Optional[Callable] = None) -> None:
```
Comment (Nic-Ma) [inline monai/data/dataset.py:42]: Can we set `Iterable` for `data` according to the doc-string?

File: monai/data/grid_dataset.py [hunk 2]
```diff
@@ -25,8 +26,13 @@ class GridPatchDataset(IterableDataset):
     """
 
     def __init__(
-        self, dataset, patch_size, start_pos=(), mode: Union[NumpyPadMode, str] = NumpyPadMode.WRAP, **pad_opts
-    ):
+        self,
+        dataset: Dataset,
+        patch_size,
```
Comment (Nic-Ma) [inline monai/data/grid_dataset.py:31]: Should patch_size be `Optional[Tuple[int, ...]]` ?

**Comment by Nic-Ma** [comment]
> could you help check this PR @Nic-Ma, for folder path string types I'm not 100% sure

Hi @wyli ，

I think @benjamin-gorman force pushed again, so I didn't find the question you said.
It overall looks good to me, could you please help double confirm?

Thanks.

**Pull request merged**